smithy-model/src/main/java/software/amazon/smithy/model/knowledge/TextInstance.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.model.knowledge;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+import java.util.Objects;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.traits.Trait;
+import software.amazon.smithy.utils.ListUtils;
+
+public final class TextInstance {
+    public enum TextLocation {
+        SHAPE,
+        APPLIED_TRAIT,
+        NAMESPACE
+    }
+
+    public enum PathElementType {
+        KEY,
+        ARRAY_INDEX
+    }
+
+    public static final class PathElement {
```

Comment:
I agree that it's the granularity of the location tracking inside trait values that adds substantial complexity to this work. I think it has value though.

If a filtered term is in an `HttpTrait`'s `uri` property value, it would direct the user to the `uri` property. That doesn't seem too valuable since the `code` or `method` properties of `HttpTrait` would not likely have a noninclusive term. However, in making no assumptions on other traits defined outside of the prelude, there could be more than one property with more free form string data, and it could be nested further.